File: app/src/main/java/suzp1984/github/io/exapidemo/app/fragments/NumFragment5Adapter.java
package suzp1984.github.io.exapidemo.app.fragments;

import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import java.util.LinkedList;
import java.util.List;

import suzp1984.github.io.exapidemo.R;

/**
 * Created by suzhenxi on 9/30/2016.
 */

public class NumFragment5Adapter extends NumFragmentBaseAdapter {

    private final String TAG = NumFragment5Adapter.class.getName();

    private final FragmentManager mFragmentManager;
    private final int COUNT = 100;
    private final List<NumberFragment> mNumFragments = new LinkedList<>();
    private final List<NumberFragment> mRemoveFragments = new LinkedList<>();

    private int mViewHolderCount = 0;

    public NumFragment5Adapter(FragmentManager manager) {
        mFragmentManager = manager;

        for (int i = 0; i < COUNT; i++) {
            mNumFragments.add(NumberFragment.newInstance(String.valueOf(i)));
        }
    }

    @Override
    public NumberViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        mViewHolderCount++;
        Log.e(TAG, "--- onCreateViewHolder: type " + viewType + ", count: " + mViewHolderCount);

        View view = LayoutInflater
                .from(parent.getContext())
                .inflate(R.layout.num_fragment_item, null);

        ViewGroup rootContainer = (ViewGroup) view;

        FrameLayout innerContainer = new FrameLayout(parent.getContext());
        innerContainer.setLayoutParams(new FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT));

        // innerContainer.setMinimumHeight(100);
        innerContainer.setId(View.generateViewId());

        rootContainer.addView(innerContainer);

        notifyViewHolderCreate(viewType);
        notifyViewHolderCountChanged(mViewHolderCount);

        return new NumFragmentBaseAdapter.NumberViewHolder(view, innerContainer);
    }

    @Override
    public void onBindViewHolder(NumberViewHolder holder, int position) {
        NumberFragment numberFragment = mNumFragments.get(position);
        holder.itemView.setTag(numberFragment);

        Log.e(TAG, "fragment: #" + numberFragment.getNumber() + " isAdded " + numberFragment.isAdded());
        Log.e(TAG, "fragment: #" + numberFragment.getNumber() + " isDetached " + numberFragment.isDetached());
        Log.e(TAG, "fragment #" + numberFragment.getNumber() + " isInLayout " + numberFragment.isInLayout());

        if (!numberFragment.isAdded()) {
            mFragmentManager.beginTransaction()
                            .replace(holder.fragmentContainer.getId(), numberFragment)
                            .commit();
        }

        if (mFragmentManager.getFragments() == null) {
            notifyAttachedFragmentChanged(0);
        } else {
            notifyAttachedFragmentChanged(mFragmentManager.getFragments().size());
        }
    }

    @Override
    public int getItemCount() {
        return mNumFragments.size();
    }

    @Override
    public void onViewRecycled(NumFragment3Adapter.NumberViewHolder viewHolder) {
        super.onViewRecycled(viewHolder);

        if (viewHolder.itemView.getTag() != null &&
                viewHolder.itemView.getTag() instanceof NumberFragment) {
            NumberFragment fragment = (NumberFragment) viewHolder.itemView.getTag();
            Log.e(TAG, "&&& onViewRecycled: " + fragment.getNumber());
            Log.e(TAG, "&&& #" + fragment.getNumber() + " isAdded " + fragment.isAdded());
            Log.e(TAG, "&&& #" + fragment.getNumber() + " isDetached " + fragment.isDetached());
            Log.e(TAG, "&&& #" + fragment.getNumber() + " isInLayout " + fragment.isInLayout());

            if (fragment.isAdded()) {
                mFragmentManager.beginTransaction().remove(fragment).commit();
            }
        }
    }

    public void resetAdapter(int size) {
        FragmentTransaction transaction = mFragmentManager.beginTransaction();
        for(NumberFragment fragment : mNumFragments) {
            if (fragment.isAdded()) {
                transaction.remove(fragment);
            }
        }

        transaction.commit();

        // mRemoveFragments.addAll(mNumFragments);

        mNumFragments.clear();

        notifyDataSetChanged();
        mViewHolderCount = 0;

        for (int i = 0; i < size; i++) {
            mNumFragments.add(NumberFragment.newInstance("new " + size + " - " + i));
        }

        notifyDataSetChanged();
    }

    public void checkFragments() {
//        mFragmentManager.popBackStack();
//        FragmentTransaction transaction = mFragmentManager.beginTransaction();
//        for (NumberFragment fragment : mRemoveFragments) {
//            Log.e(TAG, "&&& #" + fragment.getNumber() + " isAdded " + fragment.isAdded());
//            Log.e(TAG, "&&& #" + fragment.getNumber() + " isRemoving " + fragment.isRemoving());
//            Log.e(TAG, "&&& #" + fragment.getNumber() + " isInLayout " + fragment.isInLayout());
//            Log.e(TAG, "&&& #" + fragment.getNumber() + " isVisible " + fragment.isVisible());
//
//            if (fragment.isAdded()) {
//                transaction.remove(fragment);
//            }
//        }
//        transaction.commitAllowingStateLoss();

        if (mFragmentManager.getFragments() == null) {
            notifyAttachedFragmentChanged(0);
        } else {
            notifyAttachedFragmentChanged(mFragmentManager.getFragments().size());
        }

        notifyViewHolderCountChanged(mViewHolderCount);
    }
}

File: app/src/main/kotlin/suzp1984/github/io/exapidemo/app/fragments/NumFragment5Adapter.kt
package suzp1984.github.io.exapidemo.app.fragments

import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentTransaction
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout

import java.util.LinkedList

import suzp1984.github.io.exapidemo.R

/**
 * Created by suzhenxi on 9/30/2016.
 */

class NumFragment5Adapter(private val mFragmentManager: FragmentManager) : NumFragmentBaseAdapter() {

    private val TAG = NumFragment5Adapter::class.java.name
    private val COUNT = 100
    private val mNumFragments = LinkedList<NumberFragment>()

    private var mViewHolderCount = 0

    init {

        for (i in 0..COUNT - 1) {
            mNumFragments.add(NumberFragment.newInstance(i.toString()))
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NumFragmentBaseAdapter.NumberViewHolder {
        mViewHolderCount++
        Log.e(TAG, "--- onCreateViewHolder: type $viewType, count: $mViewHolderCount")

        val view = LayoutInflater
                .from(parent.context)
                .inflate(R.layout.num_fragment_item, null)

        val rootContainer = view as ViewGroup

        val innerContainer = FrameLayout(parent.context)
        innerContainer.layoutParams = FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT)

        // innerContainer.setMinimumHeight(100);
        innerContainer.id = View.generateViewId()

        rootContainer.addView(innerContainer)

        notifyViewHolderCreate(viewType)
        notifyViewHolderCountChanged(mViewHolderCount)

        return NumFragmentBaseAdapter.NumberViewHolder(view, innerContainer)
    }

    override fun onBindViewHolder(holder: NumFragmentBaseAdapter.NumberViewHolder, position: Int) {
        val numberFragment = mNumFragments[position]
        holder.itemView.tag = numberFragment

        Log.e(TAG, "fragment: #" + numberFragment.number + " isAdded " + numberFragment.isAdded)
        Log.e(TAG, "fragment: #" + numberFragment.number + " isDetached " + numberFragment.isDetached)
        Log.e(TAG, "fragment #" + numberFragment.number + " isInLayout " + numberFragment.isInLayout)

        if (!numberFragment.isAdded) {
            mFragmentManager.beginTransaction()
                    .replace(holder.fragmentContainer.id, numberFragment)
                    .commit()
        }

        if (mFragmentManager.fragments == null) {
            notifyAttachedFragmentChanged(0)
        } else {
            notifyAttachedFragmentChanged(mFragmentManager.fragments.size)
        }
    }

    override fun getItemCount(): Int {
        return mNumFragments.size
    }

    override fun onViewRecycled(viewHolder: NumFragmentBaseAdapter.NumberViewHolder?) {
        super.onViewRecycled(viewHolder)

        val f = viewHolder?.itemView?.getTag() as NumberFragment?

        if (f != null) {

            Log.e(TAG, "&&& onViewRecycled: " + f.number)
            Log.e(TAG, "&&& #" + f.number + " isAdded " + f.isAdded)
            Log.e(TAG, "&&& #" + f.number + " isDetached " + f.isDetached)
            Log.e(TAG, "&&& #" + f.number + " isInLayout " + f.isInLayout)

            if (f.isAdded) {
                mFragmentManager.beginTransaction().remove(f).commit()
            }
        }
    }

    fun resetAdapter(size: Int) {
        val transaction = mFragmentManager.beginTransaction()
        for (fragment in mNumFragments) {
            if (fragment.isAdded) {
                transaction.remove(fragment)
            }
        }

        transaction.commit()

        mNumFragments.clear()

        notifyDataSetChanged()
        mViewHolderCount = 0

        for (i in 0..size - 1) {
            mNumFragments.add(NumberFragment.newInstance("new $size - $i"))
        }

        notifyDataSetChanged()
    }

    fun checkFragments() {

        if (mFragmentManager.fragments == null) {
            notifyAttachedFragmentChanged(0)
        } else {
            notifyAttachedFragmentChanged(mFragmentManager.fragments.size)
        }

        notifyViewHolderCountChanged(mViewHolderCount)
    }
}
